File: qasdk/src/main/java/co/astrnt/qasdk/core/MyObserver.java
package co.astrnt.qasdk.core;

import com.google.gson.Gson;

import java.net.SocketTimeoutException;
import java.net.UnknownHostException;

import co.astrnt.qasdk.AstrntSDK;
import co.astrnt.qasdk.dao.BaseApiDao;
import co.astrnt.qasdk.dao.InterviewApiDao;
import co.astrnt.qasdk.dao.InterviewResultApiDao;
import co.astrnt.qasdk.dao.LogDao;
import co.astrnt.qasdk.utils.LogUtil;
import io.reactivex.Observer;
import okhttp3.ResponseBody;
import retrofit2.HttpException;

/**
 * Created by deni rohimat on 06/04/18.
 */
public abstract class MyObserver<T extends BaseApiDao> implements Observer<T> {

    protected AstrntSDK astrntSDK = new AstrntSDK();

    @Override
    public final void onComplete() {
        onApiResultCompleted();
    }

    @Override
    public final void onError(Throwable e) {
        onComplete();
        String message;
        if (e instanceof retrofit2.HttpException) {
            try {
                HttpException httpException = (HttpException) e;
                if (httpException.code() == 429) {
                    message = httpException.getMessage();
                    onApiResultError("", message, "error");
                } else {
                    ResponseBody body = ((retrofit2.HttpException) e).response().errorBody();
                    if (body != null) {
                        BaseApiDao apiDao = new Gson().fromJson(body.string(), BaseApiDao.class);
                        message = apiDao.getMessage();
                        if (apiDao.getTitle() != null) {
                            onApiResultError(apiDao.getTitle(), message, "error");
                        } else {
                            onApiResultError("", message, "error");
                        }
                    } else {
                        message = e.getMessage();
                        onApiResultError("", message, "error");
                    }
                }
            } catch (Exception e2) {
                e2.printStackTrace();
                message = e2.getMessage();
                onApiResultError("", message, "exception");
            }
        } else if (e instanceof UnknownHostException) {
            message = e.getMessage();
            onApiResultError("", message, "exception");
        } else if (e instanceof SocketTimeoutException) {
            message = e.getMessage();
            onApiResultError("", message, "exception");
        } else {
            message = e.getMessage();
            System.err.println(message);
            e.printStackTrace();
            onApiResultError("", message, "exception");
        }

        String interviewCode = astrntSDK.getInterviewCode();
        LogUtil.addNewLog(interviewCode,
                new LogDao("Response API",
                        "Response Error : " + message
                )
        );
    }

    @Override
    public final void onNext(T t) {
        onComplete();
        String interviewCode = astrntSDK.getInterviewCode();
        if (t == null) {
            onApiResultError("", "Failed connect to server", "error");

            LogUtil.addNewLog(interviewCode,
                    new LogDao("Response API",
                            "Error : Failed connect to server"
                    )
            );

            return;
        }
        if (t.getStatus() != null && t.getStatus().contains("error")) {
            if (t instanceof InterviewResultApiDao) {
                InterviewResultApiDao data = (InterviewResultApiDao) t;
                InterviewApiDao interviewApiDao = astrntSDK.getCurrentInterview();
                if (interviewApiDao != null) {
                    if (data.getToken() != null && !data.getToken().isEmpty()) {
                        LogUtil.addNewLog(interviewCode,
                                new LogDao("Token Changed",
                                        "New Token : " + data.getToken()
                                )
                        );
                        astrntSDK.saveInterview(interviewApiDao, data.getToken(), data.getInterview_code());
                    }

                    if (interviewCode != null) {
                        String message = "";
                        if (t.getMessage() != null) {
                            message = t.getMessage();
                        }
                        LogUtil.addNewLog(interviewCode,
                                new LogDao("Response API",
                                        "Error : " + message
                                )
                        );
                    }
                }
            }
            String message = "";
            if (t.getMessage() != null) {
                message = t.getMessage();
            }
            if (t.getTitle() != null) {
                onApiResultError(t.getTitle(), message, "error");
            } else {
                onApiResultError("", message, "error");
            }

            if (interviewCode != null) {
                if (t.getMessage() != null) {
                    message = t.getMessage();
                }
                LogUtil.addNewLog(interviewCode,
                        new LogDao("Response API",
                                "Error : " + message
                        )
                );
            }

        } else {
            String message = "";
            if (t.getMessage() != null) {
                message = t.getMessage();
            }
            if (interviewCode != null) {
                LogUtil.addNewLog(interviewCode,
                        new LogDao("Response API",
                                "Success : " + message
                        )
                );
            }

            onApiResultOk(t);
        }
    }

    public abstract void onApiResultCompleted();

    public abstract void onApiResultError(String title, String message, String code);

    public abstract void onApiResultOk(T t);
}

File: qasdk/src/main/java/co/astrnt/qasdk/core/MyObserver.kt
package co.astrnt.qasdk.core

import co.astrnt.qasdk.AstrntSDK
import co.astrnt.qasdk.dao.BaseApiDao
import co.astrnt.qasdk.dao.InterviewResultApiDao
import co.astrnt.qasdk.dao.LogDao
import co.astrnt.qasdk.utils.LogUtil.addNewLog
import com.google.gson.Gson
import io.reactivex.Observer
import retrofit2.HttpException
import java.net.SocketTimeoutException
import java.net.UnknownHostException


abstract class MyObserver<T : BaseApiDao> : Observer<T> {
    @JvmField var astrntSDK = AstrntSDK()
    override fun onComplete() {
        onApiResultCompleted()
    }

    override fun onError(e: Throwable) {
        onComplete()
        var message: String?
        if (e is HttpException) {
            try {
                val httpException = e
                if (httpException.code() == 429) {
                    message = httpException.message
                    onApiResultError("", message.toString(), "error")
                } else {
                    val body = e.response()!!.errorBody()
                    if (body != null) {
                        val apiDao = Gson().fromJson(body.string(), BaseApiDao::class.java)
                        message = apiDao.message
                        if (apiDao.title != null) {
                            onApiResultError(apiDao.title.toString(), message.toString(), "error")
                        } else {
                            onApiResultError("", message.toString(), "error")
                        }
                    } else {
                        message = e.message
                        onApiResultError("", message.toString(), "error")
                    }
                }
            } catch (e2: Exception) {
                e2.printStackTrace()
                message = e2.message
                onApiResultError("", message.toString(), "exception")
            }
        } else if (e is UnknownHostException) {
            message = e.message
            onApiResultError("", message.toString(), "exception")
        } else if (e is SocketTimeoutException) {
            message = e.message
            onApiResultError("", message.toString(), "exception")
        } else {
            message = e.message
            System.err.println(message)
            e.printStackTrace()
            onApiResultError("", message.toString(), "exception")
        }
        val interviewCode = astrntSDK.interviewCode
        if (message != null && message.lowercase().contains("unable to resolve host")) {
            addNewLog(interviewCode,
                    LogDao("Response API",
                            "Failed, No Internet Connection"
                    )
            )
        } else {
            addNewLog(interviewCode,
                    LogDao("Response API",
                            "Response Error : $message"
                    )
            )
        }
    }

    override fun onNext(t: T) {
        onComplete()
        val interviewCode = astrntSDK.interviewCode
        if (t == null) {
            onApiResultError("", "Failed connect to server", "error")
            addNewLog(interviewCode,
                    LogDao("Response API",
                            "Error : Failed connect to server"
                    )
            )
            return
        }
        if (t.status != null && t.status!!.contains("error")) {
            if (t is InterviewResultApiDao) {
                val data = t as InterviewResultApiDao
                val interviewApiDao = astrntSDK.currentInterview
                if (interviewApiDao != null) {
                    if (data.token != null && data.token!!.isNotEmpty()) {
                        addNewLog(interviewCode,
                                LogDao("Token Changed",
                                        "New Token : " + data.token
                                )
                        )
                        astrntSDK.saveInterview(interviewApiDao, data.token, data.interview_code)
                    }
                    if (interviewCode != null) {
                        var message: String? = ""
                        if (t.message != null) {
                            message = t.message
                        }
                        addNewLog(interviewCode,
                                LogDao("Response API",
                                        "Error : $message"
                                )
                        )
                    }
                }
            }
            var message: String? = ""
            if (t.message != null) {
                message = t.message
            }
            if (t.title != null) {
                onApiResultError(t.title.toString(), message.toString(), "error")
            } else {
                onApiResultError("", message.toString(), "error")
            }
            if (interviewCode != null) {
                if (t.message != null) {
                    message = t.message
                }
                addNewLog(interviewCode,
                        LogDao("Response API",
                                "Error : $message"
                        )
                )
            }
        } else {
            var message: String? = ""
            if (t.message != null) {
                message = t.message
            }
            if (interviewCode != null) {
                addNewLog(interviewCode,
                        LogDao("Response API",
                                "Success : $message"
                        )
                )
            }
            onApiResultOk(t)
        }
    }

    abstract fun onApiResultCompleted()
    abstract fun onApiResultError(title: String?, message: String?, code: String?)
    abstract fun onApiResultOk(baseApiDao: T)
}